File: src/prototype/web/tests/layerB.test.ts
// Unit tests for Layer B (Intent transforms).
// References docs/design/input_system_v1.md §B.1, §B.2, §B.3.

import { describe, expect, it } from "vitest";
import {
  DEFAULT_LAYER_B_CONFIG,
  INITIAL_LAYER_B_STATE,
  computeDiscreteIntents,
  computeGripIntent,
  computeHipIntent,
  transformLayerB,
  type LayerBState,
} from "../src/input/layerB.js";
import type { DiscreteIntent, GripZone } from "../src/input/intent.js";
import { ButtonBit, type InputFrame, type Vec2 } from "../src/input/types.js";

function makeFrame(overrides: Partial<InputFrame> = {}): InputFrame {
  const base: InputFrame = {
    timestamp: 0,
    ls: { x: 0, y: 0 },
    rs: { x: 0, y: 0 },
    l_trigger: 0,
    r_trigger: 0,
    buttons: 0,
    button_edges: 0,
    device_kind: "Keyboard",
  };
  return Object.freeze({ ...base, ...overrides }) as InputFrame;
}

function ls(x: number, y: number): Vec2 {
  return { x, y };
}

describe("hip intent (§B.1)", () => {
  it("centred stick produces zero hip intent", () => {
    const hip = computeHipIntent(makeFrame(), DEFAULT_LAYER_B_CONFIG);
    expect(hip.hip_angle_target).toBe(0);
    expect(hip.hip_push).toBe(0);
    expect(hip.hip_lateral).toBe(0);
  });

  it("stick pointing up yields zero angle (convention: atan2(x=0, y=+1) = 0)", () => {
    const hip = computeHipIntent(makeFrame({ ls: ls(0, 1) }), DEFAULT_LAYER_B_CONFIG);
    expect(hip.hip_angle_target).toBeCloseTo(0, 5);
    expect(hip.hip_push).toBe(1);
    expect(hip.hip_lateral).toBe(0);
  });

  it("pure-right stick yields +kAngleScale * π/2 yaw", () => {
    const hip = computeHipIntent(makeFrame({ ls: ls(1, 0) }), DEFAULT_LAYER_B_CONFIG);
    // atan2(1, 0) = π/2; scaled by 0.6.
    expect(hip.hip_angle_target).toBeCloseTo((Math.PI / 2) * 0.6, 5);
    expect(hip.hip_lateral).toBe(1);
  });

  it("hip_push and hip_lateral mirror LS exactly", () => {
    const hip = computeHipIntent(makeFrame({ ls: ls(-0.3, 0.8) }), DEFAULT_LAYER_B_CONFIG);
    expect(hip.hip_push).toBeCloseTo(0.8, 5);
    expect(hip.hip_lateral).toBeCloseTo(-0.3, 5);
  });
});

describe("grip zone selection (§B.2.1)", () => {
  it("centred stick + no triggers → no hand targets", () => {
    const { grip, nextZone } = computeGripIntent(
      makeFrame(),
      null,
      DEFAULT_LAYER_B_CONFIG,
    );
    expect(grip.l_hand_target).toBe(null);
    expect(grip.r_hand_target).toBe(null);
    expect(nextZone).toBe(null);
  });

  it("RS up-left selects COLLAR_L", () => {
    const { nextZone } = computeGripIntent(
      makeFrame({ rs: ls(-0.7, 0.7) }),
      null,
      DEFAULT_LAYER_B_CONFIG,
    );
    expect(nextZone).toBe("COLLAR_L" satisfies GripZone);
  });

  it("RS pure-down selects BELT", () => {
    const { nextZone } = computeGripIntent(
      makeFrame({ rs: ls(0, -1) }),
      null,
      DEFAULT_LAYER_B_CONFIG,
    );
    expect(nextZone).toBe("BELT" satisfies GripZone);
  });

  it("RS pure-up selects POSTURE_BREAK", () => {
    const { nextZone } = computeGripIntent(
      makeFrame({ rs: ls(0, 1) }),
      null,
      DEFAULT_LAYER_B_CONFIG,
    );
    expect(nextZone).toBe("POSTURE_BREAK" satisfies GripZone);
  });

  it("both triggers + RS aiming → both hands share the zone (§B.2.2)", () => {
    const { grip } = computeGripIntent(
      makeFrame({ rs: ls(-1, 0), l_trigger: 0.8, r_trigger: 0.5 }),
      null,
      DEFAULT_LAYER_B_CONFIG,
    );
    expect(grip.l_hand_target).toBe("WRIST_L" satisfies GripZone);
    expect(grip.r_hand_target).toBe("WRIST_L" satisfies GripZone);
    expect(grip.l_grip_strength).toBeCloseTo(0.8);
    expect(grip.r_grip_strength).toBeCloseTo(0.5);
  });

  it("only L_TRIGGER down → only left hand has a target", () => {
    const { grip } = computeGripIntent(
      makeFrame({ rs: ls(1, 0), l_trigger: 1, r_trigger: 0 }),
      null,
      DEFAULT_LAYER_B_CONFIG,
    );
    expect(grip.l_hand_target).toBe("WRIST_R" satisfies GripZone);
    expect(grip.r_hand_target).toBe(null);
  });
});

describe("grip zone hysteresis (§B.2.1)", () => {
  it("boundary nudge does NOT flip an already-selected zone", () => {
    // Held WRIST_L (pure left). Nudge RS toward up-left: direction is now
    // on the 45° boundary between WRIST_L and COLLAR_L — hysteresis should
    // keep us on WRIST_L.
    const { nextZone } = computeGripIntent(
      makeFrame({ rs: ls(-0.92, 0.38) }), // ≈ 22° above pure-left, still near WRIST_L wedge
      "WRIST_L",
      DEFAULT_LAYER_B_CONFIG,
    );
    expect(nextZone).toBe("WRIST_L" satisfies GripZone);
  });

  it("firm move past the threshold flips the zone", () => {
    // Now fully into the COLLAR_L wedge.
    const { nextZone } = computeGripIntent(
      makeFrame({ rs: ls(-0.6, 0.8) }),
      "WRIST_L",
      DEFAULT_LAYER_B_CONFIG,
    );
    expect(nextZone).toBe("COLLAR_L" satisfies GripZone);
  });

  it("releasing RS clears the zone (only when no trigger held)", () => {
    const { nextZone } = computeGripIntent(
      makeFrame({ rs: ls(0, 0) }),
      "WRIST_L",
      DEFAULT_LAYER_B_CONFIG,
    );
    expect(nextZone).toBe(null);
  });

  it("releasing RS keeps the zone if a trigger is still held", () => {
    // Player is squeezing the trigger — we should not drop the target
    // zone just because they centred the stick while gripping.
    const { nextZone } = computeGripIntent(
      makeFrame({ rs: ls(0, 0), l_trigger: 1 }),
      "COLLAR_R",
      DEFAULT_LAYER_B_CONFIG,
    );
    expect(nextZone).toBe("COLLAR_R" satisfies GripZone);
  });
});

describe("discrete intents (§B.3)", () => {
  it("L_BUMPER edge emits FOOT_HOOK_TOGGLE(L)", () => {
    const list = computeDiscreteIntents(
      makeFrame({ button_edges: ButtonBit.L_BUMPER }),
    );
    expect(list).toEqual<DiscreteIntent[]>([{ kind: "FOOT_HOOK_TOGGLE", side: "L" }]);
  });

  it("BTN_BASE held (not edge) emits BASE_HOLD every frame", () => {
    const list = computeDiscreteIntents(
      makeFrame({ buttons: ButtonBit.BTN_BASE, button_edges: 0 }),
    );
    expect(list).toEqual<DiscreteIntent[]>([{ kind: "BASE_HOLD" }]);
  });

  it("BTN_RELEASE edge emits GRIP_RELEASE_ALL", () => {
    const list = computeDiscreteIntents(
      makeFrame({ button_edges: ButtonBit.BTN_RELEASE }),
    );
    expect(list).toEqual<DiscreteIntent[]>([{ kind: "GRIP_RELEASE_ALL" }]);
  });

  it("multiple simultaneous events all surface in one list", () => {
    const list = computeDiscreteIntents(
      makeFrame({
        buttons: ButtonBit.BTN_BASE,
        button_edges:
          ButtonBit.L_BUMPER |
          ButtonBit.BTN_BREATH |
          ButtonBit.BTN_RELEASE,
      }),
    );
    expect(list).toContainEqual<DiscreteIntent>({ kind: "FOOT_HOOK_TOGGLE", side: "L" });
    expect(list).toContainEqual<DiscreteIntent>({ kind: "BASE_HOLD" });
    expect(list).toContainEqual<DiscreteIntent>({ kind: "GRIP_RELEASE_ALL" });
    expect(list).toContainEqual<DiscreteIntent>({ kind: "BREATH_START" });
    expect(list).toHaveLength(4);
  });

  it("no buttons → empty list", () => {
    expect(computeDiscreteIntents(makeFrame())).toEqual([]);
  });
});

describe("transformLayerB integration", () => {
  it("threads state across frames so hysteresis survives", () => {
    let state: LayerBState = INITIAL_LAYER_B_STATE;

    // Frame 1: aim at WRIST_L firmly.
    const r1 = transformLayerB(
      makeFrame({ rs: ls(-1, 0) }),
      state,
    );
    state = r1.nextState;
    expect(state.lastZone).toBe("WRIST_L" satisfies GripZone);

    // Frame 2: drift toward the boundary but not enough to flip.
    const r2 = transformLayerB(
      makeFrame({ rs: ls(-0.92, 0.38) }),
      state,
    );
    state = r2.nextState;
    expect(state.lastZone).toBe("WRIST_L" satisfies GripZone);

    // Frame 3: move firmly into COLLAR_L — flip.
    const r3 = transformLayerB(
      makeFrame({ rs: ls(-0.6, 0.8) }),
      state,
    );
    expect(r3.nextState.lastZone).toBe("COLLAR_L" satisfies GripZone);
  });
});

File: src/prototype/unity/Assets/BJJSimulator/Tests/EditMode/LayerBTest.cs
// Ported from src/prototype/web/tests/unit/layerB.test.ts.
// Tests for Layer B intent transforms.
// Reference: docs/design/input_system_v1.md §B.1, §B.2, §B.3.

using NUnit.Framework;

namespace BJJSimulator.Tests
{
    [TestFixture]
    public class LayerBTest
    {
        // -----------------------------------------------------------------------
        // Helpers
        // -----------------------------------------------------------------------

        static InputFrame MakeFrame(
            Vec2     ls       = default,
            Vec2     rs       = default,
            float    lTrigger = 0f,
            float    rTrigger = 0f,
            ButtonBit buttons = ButtonBit.None,
            ButtonBit edges   = ButtonBit.None) =>
            new InputFrame
            {
                Timestamp   = 0L,
                Ls          = ls,
                Rs          = rs,
                LTrigger    = lTrigger,
                RTrigger    = rTrigger,
                Buttons     = buttons,
                ButtonEdges = edges,
                DeviceKind  = DeviceKind.Keyboard,
            };

        // -----------------------------------------------------------------------
        // §B.1 — hip intent
        // -----------------------------------------------------------------------

        [Test]
        public void CentredStickProducesZeroHipIntent()
        {
            var hip = LayerBOps.ComputeHipIntent(MakeFrame(), LayerBConfig.Default);
            Assert.AreEqual(0f, hip.HipAngleTarget, 1e-6f);
            Assert.AreEqual(0f, hip.HipPush,        1e-6f);
            Assert.AreEqual(0f, hip.HipLateral,     1e-6f);
        }

        [Test]
        public void StickUpYieldsZeroAngle()
        {
            var hip = LayerBOps.ComputeHipIntent(MakeFrame(ls: new Vec2(0f, 1f)), LayerBConfig.Default);
            Assert.AreEqual(0f, hip.HipAngleTarget, 1e-5f);
            Assert.AreEqual(1f, hip.HipPush,        1e-5f);
            Assert.AreEqual(0f, hip.HipLateral,     1e-5f);
        }

        [Test]
        public void PureRightStickYieldsScaledHalfPi()
        {
            var hip = LayerBOps.ComputeHipIntent(MakeFrame(ls: new Vec2(1f, 0f)), LayerBConfig.Default);
            float expected = (float)(System.Math.PI / 2 * LayerBConfig.Default.KAngleScale);
            Assert.AreEqual(expected, hip.HipAngleTarget, 1e-5f);
            Assert.AreEqual(1f, hip.HipLateral, 1e-5f);
        }

        [Test]
        public void HipPushAndLateralMirrorLS()
        {
            var hip = LayerBOps.ComputeHipIntent(MakeFrame(ls: new Vec2(-0.3f, 0.8f)), LayerBConfig.Default);
            Assert.AreEqual( 0.8f, hip.HipPush,    1e-5f);
            Assert.AreEqual(-0.3f, hip.HipLateral, 1e-5f);
        }

        // -----------------------------------------------------------------------
        // §B.2.1 — grip zone selection
        // -----------------------------------------------------------------------

        [Test]
        public void CentredStickNoTriggersYieldsNoTargets()
        {
            var (grip, nextZone) = LayerBOps.ComputeGripIntent(MakeFrame(), GripZone.None, LayerBConfig.Default);
            Assert.AreEqual(GripZone.None, grip.LHandTarget);
            Assert.AreEqual(GripZone.None, grip.RHandTarget);
            Assert.AreEqual(GripZone.None, nextZone);
        }

        [Test]
        public void RsUpLeftSelectsCollarL()
        {
            var (_, next) = LayerBOps.ComputeGripIntent(
                MakeFrame(rs: new Vec2(-0.7f, 0.7f)), GripZone.None, LayerBConfig.Default);
            Assert.AreEqual(GripZone.CollarL, next);
        }

        [Test]
        public void RsPureDownSelectsBelt()
        {
            var (_, next) = LayerBOps.ComputeGripIntent(
                MakeFrame(rs: new Vec2(0f, -1f)), GripZone.None, LayerBConfig.Default);
            Assert.AreEqual(GripZone.Belt, next);
        }

        [Test]
        public void RsPureUpSelectsPostureBreak()
        {
            var (_, next) = LayerBOps.ComputeGripIntent(
                MakeFrame(rs: new Vec2(0f, 1f)), GripZone.None, LayerBConfig.Default);
            Assert.AreEqual(GripZone.PostureBreak, next);
        }

        [Test]
        public void BothTriggersShareSameZone()
        {
            var (grip, _) = LayerBOps.ComputeGripIntent(
                MakeFrame(rs: new Vec2(-1f, 0f), lTrigger: 0.8f, rTrigger: 0.5f),
                GripZone.None, LayerBConfig.Default);
            Assert.AreEqual(GripZone.WristL, grip.LHandTarget);
            Assert.AreEqual(GripZone.WristL, grip.RHandTarget);
            Assert.AreEqual(0.8f, grip.LGripStrength, 1e-5f);
            Assert.AreEqual(0.5f, grip.RGripStrength, 1e-5f);
        }

        [Test]
        public void OnlyLTriggerDownOnlyLeftHandTargeted()
        {
            var (grip, _) = LayerBOps.ComputeGripIntent(
                MakeFrame(rs: new Vec2(1f, 0f), lTrigger: 1f, rTrigger: 0f),
                GripZone.None, LayerBConfig.Default);
            Assert.AreEqual(GripZone.WristR, grip.LHandTarget);
            Assert.AreEqual(GripZone.None,   grip.RHandTarget);
        }

        // -----------------------------------------------------------------------
        // §B.2.1 — hysteresis
        // -----------------------------------------------------------------------

        [Test]
        public void BoundaryNudgeDoesNotFlipZone()
        {
            // Held WRIST_L; small nudge toward COLLAR_L — stays WRIST_L.
            var (_, next) = LayerBOps.ComputeGripIntent(
                MakeFrame(rs: new Vec2(-0.92f, 0.38f)),
                GripZone.WristL, LayerBConfig.Default);
            Assert.AreEqual(GripZone.WristL, next);
        }

        [Test]
        public void FirmMovePastThresholdFlipsZone()
        {
            var (_, next) = LayerBOps.ComputeGripIntent(
                MakeFrame(rs: new Vec2(-0.6f, 0.8f)),
                GripZone.WristL, LayerBConfig.Default);
            Assert.AreEqual(GripZone.CollarL, next);
        }

        [Test]
        public void ReleasingRsClearsZoneWhenNoTrigger()
        {
            var (_, next) = LayerBOps.ComputeGripIntent(
                MakeFrame(rs: Vec2.Zero), GripZone.WristL, LayerBConfig.Default);
            Assert.AreEqual(GripZone.None, next);
        }

        [Test]
        public void ReleasingRsKeepsZoneWhileTriggerHeld()
        {
            var (_, next) = LayerBOps.ComputeGripIntent(
                MakeFrame(rs: Vec2.Zero, lTrigger: 1f),
                GripZone.CollarR, LayerBConfig.Default);
            Assert.AreEqual(GripZone.CollarR, next);
        }

        // -----------------------------------------------------------------------
        // §B.3 — discrete intents
        // -----------------------------------------------------------------------

        [Test]
        public void LBumperEdgeEmitsFootHookToggleL()
        {
            var d = LayerBOps.ComputeDiscreteIntents(MakeFrame(edges: ButtonBit.LBumper));
            Assert.AreEqual(1, d.Length);
            Assert.AreEqual(DiscreteIntentKind.FootHookToggle, d[0].Kind);
            Assert.AreEqual(FootSide.L, d[0].FootSide);
        }

        [Test]
        public void BtnBaseHeldEmitsBaseHold()
        {
            var d = LayerBOps.ComputeDiscreteIntents(MakeFrame(buttons: ButtonBit.BtnBase));
            Assert.AreEqual(1, d.Length);
            Assert.AreEqual(DiscreteIntentKind.BaseHold, d[0].Kind);
        }

        [Test]
        public void BtnReleaseEdgeEmitsGripReleaseAll()
        {
            var d = LayerBOps.ComputeDiscreteIntents(MakeFrame(edges: ButtonBit.BtnRelease));
            Assert.AreEqual(1, d.Length);
            Assert.AreEqual(DiscreteIntentKind.GripReleaseAll, d[0].Kind);
        }

        [Test]
        public void MultipleSimultaneousIntentsAllSurface()
        {
            var frame = MakeFrame(
                buttons: ButtonBit.BtnBase,
                edges:   ButtonBit.LBumper | ButtonBit.BtnBreath | ButtonBit.BtnRelease);
            var d = LayerBOps.ComputeDiscreteIntents(frame);
            Assert.AreEqual(4, d.Length);

            bool hasToggle  = false, hasBase = false, hasRelease = false, hasBreath = false;
            foreach (var e in d)
            {
                if (e.Kind == DiscreteIntentKind.FootHookToggle && e.FootSide == FootSide.L) hasToggle  = true;
                if (e.Kind == DiscreteIntentKind.BaseHold)                                   hasBase    = true;
                if (e.Kind == DiscreteIntentKind.GripReleaseAll)                             hasRelease = true;
                if (e.Kind == DiscreteIntentKind.BreathStart)                                hasBreath  = true;
            }
            Assert.IsTrue(hasToggle);
            Assert.IsTrue(hasBase);
            Assert.IsTrue(hasRelease);
            Assert.IsTrue(hasBreath);
        }

        [Test]
        public void NoButtonsYieldsEmptyList()
        {
            var d = LayerBOps.ComputeDiscreteIntents(MakeFrame());
            Assert.AreEqual(0, d.Length);
        }

        // -----------------------------------------------------------------------
        // Integration: threaded state hysteresis
        // -----------------------------------------------------------------------

        [Test]
        public void TransformThreadsStateAcrossFrames()
        {
            var state = LayerBState.Initial;

            // Frame 1: aim firmly at WRIST_L.
            var (_, s1) = LayerBOps.Transform(MakeFrame(rs: new Vec2(-1f, 0f)), state);
            Assert.AreEqual(GripZone.WristL, s1.LastZone);

            // Frame 2: drift to boundary — stays WRIST_L.
            var (_, s2) = LayerBOps.Transform(MakeFrame(rs: new Vec2(-0.92f, 0.38f)), s1);
            Assert.AreEqual(GripZone.WristL, s2.LastZone);

            // Frame 3: firmly inside COLLAR_L — flips.
            var (_, s3) = LayerBOps.Transform(MakeFrame(rs: new Vec2(-0.6f, 0.8f)), s2);
            Assert.AreEqual(GripZone.CollarL, s3.LastZone);
        }
    }
}
